include/revng-c/mlir/Dialect/Clift/IR/CliftOps.td

```diff
@@ -26,6 +27,120 @@ def Clift_StatementRegion : Region<CPred<"::mlir::clift::impl::verifyRegionConte
 def Clift_ExpressionRegion : Region<CPred<"::mlir::clift::impl::verifyRegionContent<::mlir::clift::ExpressionOpInterface>($_self)">,
                                     "Region representing an expression">;
 
+//===------------------------ Top level operations ------------------------===//
+
+def Clift_ModuleOp : Clift_Op<"module",
+                              [SymbolTable,
+                               HasOnlyGraphRegion,
+                               NoRegionArguments,
+                               NoTerminator,
+                               SingleBlock,
+                               RegionKindInterface]> {
+  let regions = (region AnyRegion:$body);
+
+  let skipDefaultBuilders = 1;
+  let builders = [OpBuilder<(ins)>];
+
+  let assemblyFormat = [{
+    $body attr-dict
+  }];
+
+  let hasVerifier = 1;
+}
+
+def Clift_FunctionOp : Clift_Op<"func",
+                                [Symbol,
+                                 CallableOpInterface,
+                                 FunctionOpInterface,
+                                 AutomaticAllocationScope,
+                                 IsolatedFromAbove,
+                                 HasParent<"mlir::clift::ModuleOp">]> {
+  let arguments = (ins StrAttr:$sym_name,
```

Comment:
Also, `UniqueHandle` and `Kind` are missing. Let's discuss this tomorrow.

Reply:
We have already discussed this at lenght. Closing this thread. I'll review the implementation and comment on specific issues if relevant.